rts/lra/jaxrs/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java

```diff
@@ -460,46 +575,149 @@ private boolean isJaxRsCancel(ContainerRequestContext requestContext, ContainerR
+    // and is used for error reporting
+    private static class Progress {
+        static EnumSet<ProgressStep> failures = EnumSet.of(
+                ProgressStep.LeaveFailed,
+                ProgressStep.StartFailed,
+                ProgressStep.JoinFailed,
+                ProgressStep.CloseFailed,
+                ProgressStep.CancelFailed);
+
+        ProgressStep progress;
+        String reason;
+
+        public Progress(ProgressStep progress, String reason) {
+            this.progress = progress;
+            this.reason = reason;
+        }
+
+        public boolean ok() {
+            return !failures.contains(progress);
+        }
     }
 
-    private String getCompensatorId(URI lraId, UriInfo uriInfo, Long timeout) {
-        Map<String, String> terminateURIs = NarayanaLRAClient.getTerminationUris(resourceInfo.getResourceClass(), uriInfo, timeout);
+    // convert the list of steps carried out by the filters into a warning message
+    private String processLRAOperationFailures(ArrayList<Progress> progress) {
+        StringJoiner badOps = new StringJoiner(", ");
+        StringJoiner goodOps = new StringJoiner(", ");
+        StringBuffer code = new StringBuffer("-");
 
-        if (!terminateURIs.containsKey("Link")) {
-            throwGenericLRAException(lraId, Response.Status.BAD_REQUEST.getStatusCode(),
-                    "Missing complete or compensate annotations");
-        }
+        progress.forEach(p -> {
+            if (p.ok()) {
+                code.insert(0, p.progress.ordinal());
+                goodOps.add(String.format("%s (%s)", p.progress.name(), p.progress.status));
+            } else {
+                code.append(p.progress.ordinal());
+                badOps.add(String.format("%s (%s)", p.progress.name(), p.reason));
+            }
+        });
+
+        /*
+         * return a string which encodes the result:
+         * <major code>-<failed op codes>-<successful op codes>: <details of failed ops> (<details of successful ops>)
+         *
+         * where
+         *
+         * <major code>: corresponds to the id of the message in the logs
+         * <failed op codes>: each digit corresponds to the enum ordinal calue of the ProgressStep enum value that was successful
+         * <successful op codes>: each digit corresponds to the enum ordinal calue of the ProgressStep enum value that failed
+         * <details of failed ops>: comma separated list of failed operation details "<op name> (<exception message>)"
+         * <details of successful ops>: comma separated list of successful operation details "<op name> (<op description>)"
+         */
+
+        return badOps.length() != 0 ? String.format("%d-%s: %s (%s)", i18nMessageCode, code, badOps, goodOps) : null;
+    }
 
-        return terminateURIs.get("Link");
+    private boolean progressDoesNotContain(ArrayList<Progress> progress, ProgressStep step) {
+        return progress.stream().noneMatch(p -> p.progress == step);
```

Comment:
When the progress data structure is used for seeking on existence would not be better to use the map structure?

Reply:
I don't know. Let's get the functionality out there rather than getting the code absolutely perfect. There is plenty of scope for improving the code elsewhere.